evm/src/fuzz/invariant/error.rs

```diff
@@ -0,0 +1,224 @@
+            ));
+
+            counterexample_sequence.push(BaseCounterExample::create(
+                *sender,
+                *addr,
+                bytes,
+                &ided_contracts,
+            ));
+
+            // Checks the invariant.
+            if let Some(func) = &self.func {
+                let error_call_result = executor
+                    .call_raw(CALLER, self.addr, func.0.clone(), 0.into())
+                    .expect("bad call to evm");
+
+                if error_call_result.reverted {
+                    logs.extend(error_call_result.logs);
+                    traces.push((TraceKind::Execution, error_call_result.traces.unwrap()));
+                    break
+                }
+            }
+        }
+
+        (!counterexample_sequence.is_empty())
+            .then_some(CounterExample::Sequence(counterexample_sequence))
+    }
+
+    /// Tests that the sequence of calls successfully reverts on the error function.
+    fn fails_successfully(&self, mut executor: Executor, calls: &[&BasicTxDetails]) -> bool {
+        for (sender, (addr, bytes)) in calls {
+            executor
+                .call_raw_committing(*sender, *addr, bytes.0.clone(), 0.into())
+                .expect("bad call to evm");
+
+            // Checks the invariant.
+            if let Some(func) = &self.func {
+                let error_call_result = executor
+                    .call_raw(CALLER, self.addr, func.0.clone(), 0.into())
+                    .expect("bad call to evm");
+
+                if error_call_result.reverted {
+                    return true
+                }
+            }
+        }
+        false
+    }
+
+    /// Tries to shrink the failure case to its smallest sequence of calls.
+    ///
+    /// Sets an anchor at the beginning (index=0) and tries to remove all other calls one by one,
+    /// until it reaches the last one. The elements which were removed and lead to a failure are
+    /// kept in the removal list. The removed ones that didn't lead to a failure are inserted
+    /// back into the sequence.
+    ///
+    /// Once it reaches the end, it increments the anchor, resets the removal list and starts the
+    /// same process again.
+    ///
+    /// Returns the smallest sequence found.
+    fn try_shrinking<'a>(
+        &self,
+        calls: &'a [BasicTxDetails],
+        executor: &Executor,
+    ) -> Vec<&'a BasicTxDetails> {
+        let mut anchor = 0;
+        let mut removed_calls = vec![];
+        let mut shrinked = calls.iter().collect::<Vec<_>>();
+
+        while anchor != calls.len() {
+            let new_sequence = calls
+                .iter()
+                .enumerate()
+                .filter_map(|(index, element)| {
+                    if anchor > index || removed_calls.contains(&index) {
+                        return None
+                    }
+                    Some(element)
+                })
+                .collect::<Vec<_>>();
```

Comment:
nit: you don't need to collect here given you'll iterate over this immutably again in `fails_successfully`